Update dvc root command reference motivation and examples (#1637)
iterative/dvc.org: content/docs/command-reference/root.md

## Description

> Returns the path to the root directory of the <abbr>DVC project</abbr>, relative

Reviewer: Should we also say "root directory" in the short intro (line 3) ?

Author: Agreed. Added this.

Reviewer: Forgot to push?

Author: Was waiting for comments on other parts. Will push this by itself for now.

> to the current working directory. Useful when working in a subdirectory of the

Reviewer: Sorry I'm going to be annoying and since I can't un-see your previous version, I think at least mentioning the command name once here would be nice:
```suggestion
to the current working directory. `dvc root` is useful when working in a
subdirectory of the ...
```
> Plus it makes it a complete sentence.

Reviewer: Or maybe "`dvc repro` is useful to refer to a file in another directory when working in a subdirectory of the project."

Author: Agree with the change, but I'll wait on team comments before updating motivation.

Author: Adjusted this but used the `dvc root` reference in the second paragraph instead.

project, and you need to refer to a file in another directory. Use it in files

> and commands to build a path to a dependency, script, or <abbr>data

Reviewer: > Use it in files and commands

How do you use a command in files and commands? I'm not sure what we mean here 🙂

Author: Yes, this part is confusing, probably because there are effectively two usages: `dvc root` the command and `$(dvc root)` used in another command or script. I think this will make more sense and can be rewritten once clarify the recommended usage.

Reviewer: And probably worth it's own paragraph after all. I did come to conclusions in https://github.com/iterative/dvc.org/issues/404#issuecomment-666104957 but feel free to wait a bit and see if the team has more feedback. Thanks!

Author: Split to second paragraph and adjusted language — instead of "files and commands" the primary motivation is "build reusable paths" as per discussion on #404.

Reviewer: OK, better, thanks. Resolving here but I have more feedback on this...

## Description

Returns the path to the root directory of the <abbr>DVC project</abbr>, relative

> to the current working directory. Useful for referencing a file in another
> directory when working in a subdirectory of the project.

Reviewer: ```suggestion
to the current working directory. `dvc root` is useful for referencing a file
in another directory when working in a subdirectory of the project.
```

> Use `dvc root` to build reusable paths to dependencies, scripts, or <abbr>data
> artifacts</abbr>.

Reviewer: ```suggestion
Use this command to build reusable paths to dependencies, scripts, or
<abbr>data artifacts</abbr>.
```

> <abbr>data artifacts</abbr>.

Reviewer: What is a "reusable path"? Still needs more explanation on the command motivation or main use case(s) as discussed in #404. It's OK if the paragraph grows (in fact I was expecting so, thus recommended separating this into its own p).

Author: OK, working on expanding this to include scripting motivation.

Author: Expanded this with another sentence. The POSIX use case seems platform-specific so I left it out.

## Description

Returns the path to the root directory of the <abbr>DVC project</abbr>, relative
to the current working directory. `dvc root` is useful for referencing a file in
another directory when working in a subdirectory of the project.

Use this command to build fixed paths to dependencies, files, or <abbr>data

> artifacts</abbr>. Useful when creating shell scripts that generate `dvc.yaml`
> files programmatically or with the `dvc run` helper.

Reviewer: ```suggestion
artifacts</abbr>. This is useful when creating shell scripts that generate
`dvc.yaml` files programmatically or with the `dvc run` helper.
```

## Description

> The `dvc root` command returns the path to the root directory of the <abbr>DVC

Reviewer: Let's keep it short :)
```suggestion
Returns the path to the root directory of the <abbr>DVC
```

Your version is correct but we try to avoid repetition and especially for `dvc {cmd}` because those get linked automatically to the cmd ref. We don't need so many links back to the same web page.

Reviewer: This change will require reformatting the paragraph.

Author: Removed extra links and shortened first sentence as suggested.

Reviewer: Thanks. Could've left one instance 😅 but it's OK either way.

project</abbr>, relative to the current working directory.

> Using `dvc root`, you can build a path to a dependency, script, or <abbr>data

Reviewer: Why were the sentences separated into a paragraph each? I think they're pretty closely related, at least these 2 first ones.

Author: Returned to single paragraph format.

Reviewer: Thanks. My Q wasn't necessarily rethorical BTW, you're more than welcome to insist on changes if you feel stringly (just FYI). Resolving

artifact</abbr> using `dvc run` from any subdirectory of a <abbr>DVC

> project</abbr>.

Reviewer: >build a path ... using `dvc run` from any subdirectory of a DVC project.

I don't understand this.

Author: Edited this sentence to simplify.

You can also use `dvc root` to simplify other commands when working in a

> <abbr>DVC project</abbr>.

Reviewer: I don't see much meaning in this sentence.

The old one "Useful when working in a subdirectory of the project, and needing to refer to a file in another directory." explained the main use case.

Author: Returned this previous version with a minor edit for "needing to".

```dvc
$ dvc root
.

$ cd subdir
$ dvc root
..
```

> ## Example: Referencing files

Reviewer: Maybe something like
```suggestion
## Example: Referencing fixed paths for `dvc run`
```
?

Simplify file references when working in a subdirectory of a <abbr>DVC

> project</abbr>.

Reviewer: Nice use of abbr 👍 

Maybe add something like "...by making all paths relative to the root dir?" to explain what you mean by "simplify"?

Let's be explicit 🙂

Author: Updated this and made consistent between examples.

Reviewer: Looks good, thanks.

```dvc
$ dvc root
../../../

$ dvc run -d $(dvc root)/data/file.cvs ... \
    python $(dvc root)/scripts/something.py
```

## Example: Output references

Simplify output file or directory references.

```dvc
$ dvc root
../../../

$ dvc get -o $(dvc root)/root-model.pkl \
    https://github.com/iterative/example-get-started model.pkl

> ```

Reviewer: Thanks but seems to me like the same use case as the previous example, and probably `run` is much more realistic. Doubt people will need this for `get`.

Author: > > I adjusted the second example... uses dvc root to specify the output of a dvc get command...
> 
> I don't see this in the 2nd example.

Sorry, I meant to refer to this third example. 

I agree the use cases are similar. I used `get` because it doesn't rely on how `dvc run` executes or how stages are generated. As someone new to dvc, this seemed more straightforward to understanding how `dvc root` works on its own.

I think that examples should get progressively more complex but was hesitant to move or replace the existing second example on the first pass.

Reviewer: Yes, its good thining. But tbh `dvc root` is far from a well-known or frequently-used command so we don't need so many examples. The minimum possible is best here.

Author: Removed this example.

Reviewer: Thanks

## Example: Other commands

> Simplify other commands when working in a <abbr>DVC project</abbr>.

Reviewer: OK interesting example. Let's put it all the way at the end please. Also could the title be a little more specific, as well as explaining what we mean by "simplify" (similar to my previous comment).

Author: Moved to the end and updated title to be more specific.

Reviewer: The title ended up the same as what I changed the 2nd example to haha. So I merged them.

```dvc
$ dvc root
../..

$ tree $(dvc root)/data/
../../data/
├── data.xml
├── data.xml.dvc
...
└── prepared
    ├── test.tsv
    └── train.tsv
```

## Example: Build reusable paths

Build reusable paths to dependencies, scripts, or <abbr>data artifacts</abbr>
from separate stages and subdirectories.

```dvc
$ cd more_stages/
$ dvc run -n process_data \
          -d data.in \

>           -d $(dvc root)/process_data.py \

Reviewer: OK... I think what we need is a shell script that generates a dvc.yaml file instead of another `dvc run` example. Not required for this PR, so if you're not so familiar with Linux shell just leave it out for now please. Thanks

Author: I removed this for now. 

I looked into some related issues (#615) and [examples of scripts](https://github.com/bertbesser/dvc-walkthrough/blob/master/scripts/walkthrough.sh). 

These make sense within the scripting use case but instead of adding a (complicated) example that builds out a full set of stages, it would help to have the team vet a smaller example that fits the recommended usage.

```dvc
$ dvc root
.
$ mkdir subdir
$ cd subdir
$ dvc root
..
```

## Example: Referencing fixed paths in other commands

When working in a subdirectory of a <abbr>DVC project</abbr>, simplify file
references by making all paths relative to the root directory.

```dvc
$ dvc root
../../../

$ dvc run -d $(dvc root)/data/file.cvs ... \
    python $(dvc root)/scripts/something.py
```

> ## Example: Referencing fixed paths for other commands
> 
> When working in a subdirectory of a <abbr>DVC project</abbr>, simplify file
> references in other commands by using paths relative to the root directory.

Reviewer: Let's merge these 2 examples:
```suggestion
The same can be applied to non-DVC commands:
```

```dvc
$ dvc root
../..

$ tree $(dvc root)/data/
../../data/
├── data.xml
├── data.xml.dvc
...

> └── prepared
>     ├── test.tsv
>     └── train.tsv

Reviewer: Shortening
```suggestion
```

```dvc
$ dvc root
.

> 

Reviewer: ```suggestion
```

$ mkdir subdir
$ cd subdir
$ dvc root
..
```

> ## Example: Referencing fixed paths for `dvc run`

Reviewer: ```suggestion
## Example: Referencing fixed paths in other commands
```

## Example: Referencing files

Simplify file references when working in a subdirectory of a <abbr>DVC
project</abbr>.

```dvc
$ dvc root

> 

Reviewer: ```suggestion
```
Let's also remove all other empty lines in example code blocks, at least after dvc root (the actual output has no extra line).

Author: Did this and made the last one consistent, adding a space before the `$ tree` command begins.

Reviewer: Thanks
